File: vanguard/packages/domain/contracts.ts
import { createHash } from "node:crypto";
import path from "node:path";

export type Json = null | boolean | number | string | Json[] | { [key: string]: Json };
export type Digest = `sha256:${string}`;

export type ResourceSelector =
  | { kind: "fs"; root: string; paths: string[] }
  | { kind: "network"; hosts: string[]; ports: number[] }
  | { kind: "secret"; refs: string[]; discloseToModel: false }
  | { kind: "git"; repo: string; refs: string[] }
  | { kind: "table"; table: string; ranges?: string[] }
  | { kind: "browser"; origin: string; accountRef?: string }
  | { kind: "generic"; uriPattern: string };

export interface EffectDescriptor {
  name: string;
  args: Record<string, Json>;
  digest: Digest;
  readSet?: ResourceSelector[];
  writeSet?: ResourceSelector[];
}

export interface CapabilityGrant {
  id: string;
  principal: string;
  descriptorDigest: Digest;
  actions: string[];
  resources: ResourceSelector[];
  constraints: {
    expiresAt: string;
    maxUses: string;
    maxBytes?: string;
    maxEffects?: string;
    budgetLeaseId: string;
    environmentSnapshot?: Digest;
    networkPolicy?: "deny" | "allowlist";
    requirePreview?: boolean;
    requireApprovalAboveRisk?: "low" | "medium" | "high" | "critical";
  };
  purposeDigest: Digest;
  parentGrantId?: string;
  approvalRef?: string;
  authenticator?: string;
}

export interface AffectedResource {
  resource: string;
  change: "created" | "modified" | "deleted" | "observed";
  preDigest?: Digest;
  postDigest?: Digest;
  patchRef?: Digest;
}

export interface Receipt {
  descriptorDigest: Digest;
  grantId?: string;
  outcome: "ok" | "failed" | "undeterminable";
  observedAt: string;
  resultDigest: Digest;
  affectedResources: AffectedResource[];
  workingDirectory?: string;
  uncertainty?: { scope: "effect_occurrence" | "evidence_completeness" | "result"; reason: string };
  note?: string;
}

export interface EventEnvelope {
  schemaVersion: "vg.4";
  eventId: string;
  scope: "episode" | "governance" | "evolution" | "recovery";
  runId?: string;
  episodeId?: string;
  branchId?: number;
  parentEventId?: string;
  traceId?: string;
  spanId?: string;
  seq: string;
  occurredAt: string;
  recordedAt: string;
  principal: string;
  tenantId: string;
  ownerId: string;
  confidentiality: "public" | "internal" | "confidential" | "restricted";
  retentionClass: "ephemeral" | "standard" | "extended" | "legal_hold";
  trainability: "prohibited" | "opt_in_required" | "opt_in_granted";
  redactionStatus: "none" | "partial" | "complete" | "pending";
  encryptionKeyRef?: string;
  environmentSnapshot?: Digest;
  payload: { kind: string; [key: string]: Json };
}

export interface InvalidationCondition {
  condition: string;
  checkKind: "automatic" | "scheduled" | "manual";
  checkRef?: string;
}

export interface Artifact {
  id: string;
  kind: "R" | "O" | "M" | "P";
  artifactVersion: string;
  body: Digest;
  interfaceSchema: string;
  createdBy: string;
  createdFrom: string[];
  dependencies: { artifact: string; versionRange: string }[];
  supersedes: string[];
  contentDigest: Digest;
  createdAt: string;
  invalidationConditions: InvalidationCondition[];
}

export interface EvidenceClaim {
  id: string;
  subject: string;
  predicate: string;
  value: Json;
  protocol: Digest;
  evaluator: {
    evaluatorId: string;
    class: "mechanically_reproducible" | "externally_grounded" | "human_adjudicated" | "learned_proxy" | "composite" | "inconclusive";
    imageDigest: Digest;
  };
  environmentProfile: Digest;
  substrateProfile: Digest;
  taskDistribution: Digest;
  uncertainty: { kind: "interval" | "point" | "qualitative" | "unknown"; lower?: number; upper?: number; n?: number; note?: string };
  validity: { domains: string[]; note?: string };
  invalidationConditions: InvalidationCondition[];
  evidenceRefs?: Digest[];
  derivedFrom?: string[];
  contradicts?: string[];
  expiresAt?: string;
}

const DIGEST = /^sha256:[0-9a-f]{64}$/;
const TIMESTAMP = /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d{3}Z$/;
const UUID_V7 = /^[0-9a-f]{8}-[0-9a-f]{4}-7[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/;
const INT_STRING = /^(0|[1-9][0-9]*)$/;

function fail(message: string): never { throw new TypeError(message); }
function object(value: unknown, label: string): Record<string, unknown> {
  if (value === null || typeof value !== "object" || Array.isArray(value)) fail(`${label} must be an object`);
  return value as Record<string, unknown>;
}
function string(value: unknown, label: string): string {
  if (typeof value !== "string" || value.length === 0) fail(`${label} must be a non-empty string`);
  return value;
}
function digest(value: unknown, label: string): Digest {
  const result = string(value, label);
  if (!DIGEST.test(result)) fail(`${label} must be a sha256 digest`);
  return result as Digest;
}
function timestamp(value: unknown, label: string): string {
  const result = string(value, label);
  if (!TIMESTAMP.test(result)) fail(`${label} must be an RFC 3339 UTC timestamp with millisecond precision`);
  return result;
}
function nonEmptyArray(value: unknown, label: string): unknown[] {
  if (!Array.isArray(value) || value.length === 0) fail(`${label} must be a non-empty array`);
  return value;
}
function oneOf<T extends string>(value: unknown, values: readonly T[], label: string): T {
  if (typeof value !== "string" || !values.includes(value as T)) fail(`${label} has an invalid value`);
  return value as T;
}

/** RFC 8785-compatible for JSON-domain values (including sorted object keys). */
export function canonicalJson(value: Json): string {
  if (value === null || typeof value !== "object") {
    const encoded = JSON.stringify(value);
    if (encoded === undefined || (typeof value === "number" && !Number.isFinite(value))) fail("value is not canonicalisable JSON");
    return encoded;
  }
  if (Array.isArray(value)) return `[${value.map(canonicalJson).join(",")}]`;
  return `{${Object.keys(value).sort().map((key) => `${JSON.stringify(key)}:${canonicalJson(value[key]!)}`).join(",")}}`;
}

export function sha256(value: Json): Digest {
  return `sha256:${createHash("sha256").update(canonicalJson(value), "utf8").digest("hex")}`;
}

function normalizePath(value: string, workspaceRoot: string): string {
  const root = path.posix.resolve("/", workspaceRoot.replaceAll("\\", "/"));
  return path.posix.resolve(root, value.replaceAll("\\", "/"));
}

export function createEffectDescriptor(
  name: string,
  args: Record<string, Json | undefined>,
  options: { workspaceRoot: string; pathArguments?: readonly string[]; readSet?: ResourceSelector[]; writeSet?: ResourceSelector[] },
): EffectDescriptor {
  string(name, "EffectDescriptor.name");
  const pathArguments = new Set(options.pathArguments ?? []);
  const normalized: Record<string, Json> = {};
  for (const [key, value] of Object.entries(args)) {
    if (value === undefined) continue;
    normalized[key] = pathArguments.has(key) && typeof value === "string" ? normalizePath(value, options.workspaceRoot) : value;
  }
  const result: EffectDescriptor = { name, args: normalized, digest: sha256({ name, args: normalized }) };
  if (options.readSet !== undefined) result.readSet = options.readSet;
  if (options.writeSet !== undefined) result.writeSet = options.writeSet;
  return result;
}

export function parseEffectDescriptor(value: unknown): EffectDescriptor {
  const input = object(value, "EffectDescriptor");
  const name = string(input.name, "EffectDescriptor.name");
  const args = object(input.args, "EffectDescriptor.args") as Record<string, Json>;
  const supplied = digest(input.digest, "EffectDescriptor.digest");
  if (sha256({ name, args }) !== supplied) fail("EffectDescriptor.digest does not bind name and args");
  return structuredClone(input) as unknown as EffectDescriptor;
}

export function parseCapabilityGrant(value: unknown): CapabilityGrant {
  const input = object(value, "CapabilityGrant");
  string(input.id, "CapabilityGrant.id");
  string(input.principal, "CapabilityGrant.principal");
  digest(input.descriptorDigest, "CapabilityGrant.descriptorDigest");
  nonEmptyArray(input.actions, "CapabilityGrant.actions").forEach((v) => string(v, "CapabilityGrant.actions[]"));
  nonEmptyArray(input.resources, "CapabilityGrant.resources");
  const constraints = object(input.constraints, "CapabilityGrant.constraints");
  timestamp(constraints.expiresAt, "CapabilityGrant.constraints.expiresAt");
  if (!INT_STRING.test(string(constraints.maxUses, "CapabilityGrant.constraints.maxUses"))) fail("maxUses must be an IntString");
  string(constraints.budgetLeaseId, "CapabilityGrant.constraints.budgetLeaseId");
  digest(input.purposeDigest, "CapabilityGrant.purposeDigest");
  return structuredClone(input) as unknown as CapabilityGrant;
}

function parseInvalidation(value: unknown, label: string): void {
  nonEmptyArray(value, label).forEach((entry, index) => {
    const condition = object(entry, `${label}[${index}]`);
    string(condition.condition, `${label}[${index}].condition`);
    const kind = oneOf(condition.checkKind, ["automatic", "scheduled", "manual"] as const, `${label}[${index}].checkKind`);
    if (kind === "automatic") string(condition.checkRef, `${label}[${index}].checkRef`);
  });
}

export function parseReceipt(value: unknown): Receipt {
  const input = object(value, "Receipt");
  digest(input.descriptorDigest, "Receipt.descriptorDigest");
  oneOf(input.outcome, ["ok", "failed", "undeterminable"] as const, "Receipt.outcome");
  timestamp(input.observedAt, "Receipt.observedAt");
  digest(input.resultDigest, "Receipt.resultDigest");
  if (!Array.isArray(input.affectedResources)) fail("Receipt.affectedResources must be an array");
  input.affectedResources.forEach((entry, index) => {
    const affected = object(entry, `Receipt.affectedResources[${index}]`);
    string(affected.resource, `Receipt.affectedResources[${index}].resource`);
    const change = oneOf(affected.change, ["created", "modified", "deleted", "observed"] as const, `Receipt.affectedResources[${index}].change`);
    if (change !== "created" && affected.preDigest === undefined) fail(`${change} resource requires preDigest`);
    if (change !== "deleted" && affected.postDigest === undefined) fail(`${change} resource requires postDigest`);
    if (affected.preDigest !== undefined) digest(affected.preDigest, "preDigest");
    if (affected.postDigest !== undefined) digest(affected.postDigest, "postDigest");
  });
  if (input.outcome === "undeterminable" && input.uncertainty === undefined) fail("undeterminable Receipt requires uncertainty");
  if (input.uncertainty !== undefined) {
    const uncertainty = object(input.uncertainty, "Receipt.uncertainty");
    oneOf(uncertainty.scope, ["effect_occurrence", "evidence_completeness", "result"] as const, "Receipt.uncertainty.scope");
    string(uncertainty.reason, "Receipt.uncertainty.reason");
  }
  return structuredClone(input) as unknown as Receipt;
}

export function parseEventEnvelope(value: unknown): EventEnvelope {
  const input = object(value, "EventEnvelope");
  if (input.schemaVersion !== "vg.4") fail("EventEnvelope.schemaVersion must be vg.4");
  if (!UUID_V7.test(string(input.eventId, "EventEnvelope.eventId"))) fail("EventEnvelope.eventId must be UUIDv7");
  const scope = oneOf(input.scope, ["episode", "governance", "evolution", "recovery"] as const, "EventEnvelope.scope");
  if ((scope === "episode" || scope === "recovery") && input.runId === undefined) fail(`${scope} EventEnvelope requires runId`);
  if (scope === "episode" && input.episodeId === undefined) fail("episode EventEnvelope requires episodeId");
  if (!INT_STRING.test(string(input.seq, "EventEnvelope.seq"))) fail("EventEnvelope.seq must be an IntString");
  timestamp(input.occurredAt, "EventEnvelope.occurredAt");
  timestamp(input.recordedAt, "EventEnvelope.recordedAt");
  ["principal", "tenantId", "ownerId"].forEach((key) => string(input[key], `EventEnvelope.${key}`));
  string(object(input.payload, "EventEnvelope.payload").kind, "EventEnvelope.payload.kind");
  return structuredClone(input) as unknown as EventEnvelope;
}

export function parseArtifact(value: unknown): Artifact {
  const input = object(value, "Artifact");
  string(input.id, "Artifact.id");
  oneOf(input.kind, ["R", "O", "M", "P"] as const, "Artifact.kind");
  string(input.artifactVersion, "Artifact.artifactVersion");
  digest(input.body, "Artifact.body");
  digest(input.contentDigest, "Artifact.contentDigest");
  timestamp(input.createdAt, "Artifact.createdAt");
  parseInvalidation(input.invalidationConditions, "Artifact.invalidationConditions");
  return structuredClone(input) as unknown as Artifact;
}

export function parseEvidenceClaim(value: unknown): EvidenceClaim {
  const input = object(value, "EvidenceClaim");
  ["id", "subject", "predicate"].forEach((key) => string(input[key], `EvidenceClaim.${key}`));
  ["protocol", "environmentProfile", "substrateProfile", "taskDistribution"].forEach((key) => digest(input[key], `EvidenceClaim.${key}`));
  const evaluator = object(input.evaluator, "EvidenceClaim.evaluator");
  string(evaluator.evaluatorId, "EvidenceClaim.evaluator.evaluatorId");
  digest(evaluator.imageDigest, "EvidenceClaim.evaluator.imageDigest");
  nonEmptyArray(object(input.validity, "EvidenceClaim.validity").domains, "EvidenceClaim.validity.domains");
  parseInvalidation(input.invalidationConditions, "EvidenceClaim.invalidationConditions");
  return structuredClone(input) as unknown as EvidenceClaim;
}

export function serialize<T>(value: T): string { return JSON.stringify(value); }
export function deserialize<T>(wire: string, parser: (value: unknown) => T): T { return parser(JSON.parse(wire)); }

File: vanguard/packages/domain/wire/contracts.py
"""Reader-profile implementation for the T1.4–T1.12 wire contracts.

Readers validate every known field and preserve unknown fields recursively.
Writer strictness remains normative in ``schemas/v4/*.schema.json``.
"""

from __future__ import annotations

from copy import deepcopy
from typing import Any, Callable, Mapping

from ..canonicalisation.digest import digest_of
from ..primitives.primitives import ParseError, parse
from ..selectors.resource_selector import SELECTOR_KINDS, SelectorError, parse_selector

JsonObject = dict[str, Any]


class WireError(ValueError):
    def __init__(self, code: str, path: str, message: str) -> None:
        super().__init__(f"{path}: {message} ({code})")
        self.code = code
        self.path = path


def _object(value: Any, path: str) -> JsonObject:
    if not isinstance(value, dict):
        raise WireError("type", path, "expected object")
    return value


def _array(value: Any, path: str, *, nonempty: bool = False) -> list[Any]:
    if not isinstance(value, list):
        raise WireError("type", path, "expected array")
    if nonempty and not value:
        raise WireError("minItems", path, "must not be empty")
    return value


def _string(value: Any, path: str) -> str:
    if not isinstance(value, str):
        raise WireError("type", path, "expected string")
    if not value:
        raise WireError("minLength", path, "must not be empty")
    return value


def _integer(value: Any, path: str) -> int:
    if isinstance(value, bool) or not isinstance(value, int):
        raise WireError("type", path, "expected integer")
    return value


def _required(value: Mapping[str, Any], fields: tuple[str, ...], path: str) -> None:
    for field in fields:
        if field not in value:
            raise WireError("required", path, f"missing {field}")


def _enum(value: Any, options: tuple[str, ...], path: str) -> str:
    result = _string(value, path)
    if result not in options:
        raise WireError("enum", path, f"unsupported value {result!r}")
    return result


def _primitive(kind: str, value: Any, path: str) -> None:
    try:
        parse(kind, value)
    except ParseError as exc:
        raise WireError(exc.code, path, str(exc)) from exc


def _strings(value: Any, path: str, *, nonempty: bool = False) -> list[str]:
    items = _array(value, path, nonempty=nonempty)
    result = [_string(item, f"{path}/{index}") for index, item in enumerate(items)]
    if len(result) != len(set(result)):
        raise WireError("uniqueItems", path, "items must be unique")
    return result


def _selector(value: Any, path: str) -> None:
    source = _object(value, path)
    kind = source.get("kind")
    if kind not in SELECTOR_KINDS:
        raise WireError("enum", f"{path}/kind", "unknown selector kind")
    known_fields = {
        "fs": ("kind", "root", "paths"),
        "network": ("kind", "hosts", "ports"),
        "secret": ("kind", "refs", "discloseToModel"),
        "git": ("kind", "repo", "refs"),
        "table": ("kind", "table", "ranges"),
        "browser": ("kind", "origin", "accountRef"),
        "generic": ("kind", "uriPattern"),
    }[kind]
    candidate = {field: source[field] for field in known_fields if field in source}
    try:
        parse_selector(candidate)
    except SelectorError as exc:
        raise WireError(exc.code, path, str(exc)) from exc


def _provenance(value: Any, path: str) -> None:
    source = _object(value, path)
    _required(source, ("origin", "instructionAuthority", "integrity", "confidentiality", "epistemic"), path)
    _enum(source["origin"], ("operator", "system", "environment", "model", "external", "memory"), f"{path}/origin")
    _enum(source["instructionAuthority"], ("directive", "advisory", "inert"), f"{path}/instructionAuthority")
    _enum(source["integrity"], ("attested", "verified", "unverified", "tampered"), f"{path}/integrity")
    _primitive("ConfidentialityLabel", source["confidentiality"], f"{path}/confidentiality")
    _primitive("EpistemicState", source["epistemic"], f"{path}/epistemic")


def _invalidation(value: Any, path: str) -> None:
    for index, item in enumerate(_array(value, path, nonempty=True)):
        item_path = f"{path}/{index}"
        source = _object(item, item_path)
        _required(source, ("condition", "checkKind"), item_path)
        _string(source["condition"], f"{item_path}/condition")
        check_kind = _enum(source["checkKind"], ("automatic", "scheduled", "manual"), f"{item_path}/checkKind")
        if check_kind == "automatic":
            if "checkRef" not in source:
                raise WireError("required", item_path, "automatic condition requires checkRef")
            _primitive("EvaluatorId", source["checkRef"], f"{item_path}/checkRef")


def _parse_effect(value: Any) -> None:
    source = _object(value, "EffectDescriptor")
    _required(source, ("verb", "sinkClass", "selector", "args", "argsDigest", "idempotencyKey", "riskTier", "provenance"), "EffectDescriptor")
    _string(source["verb"], "EffectDescriptor/verb")
    _enum(source["sinkClass"], ("pure", "observation", "privileged"), "EffectDescriptor/sinkClass")
    _selector(source["selector"], "EffectDescriptor/selector")
    args = _object(source["args"], "EffectDescriptor/args")
    _primitive("Digest", source["argsDigest"], "EffectDescriptor/argsDigest")
    if digest_of(args) != source["argsDigest"]:
        raise WireError("digest", "EffectDescriptor/argsDigest", "does not bind canonical args")
    if "workingDirectory" in source:
        _string(source["workingDirectory"], "EffectDescriptor/workingDirectory")
    for field in ("readSet", "writeSet"):
        if field in source:
            for index, selector in enumerate(_array(source[field], f"EffectDescriptor/{field}")):
                _selector(selector, f"EffectDescriptor/{field}/{index}")
    _string(source["idempotencyKey"], "EffectDescriptor/idempotencyKey")
    _primitive("RiskTier", source["riskTier"], "EffectDescriptor/riskTier")
    _provenance(source["provenance"], "EffectDescriptor/provenance")


def _parse_grant(value: Any) -> None:
    source = _object(value, "CapabilityGrant")
    _required(source, ("grantId", "principal", "descriptorDigest", "actions", "selector", "constraints", "expiry", "maxUses", "purposeDigest"), "CapabilityGrant")
    for field, kind in (("grantId", "GrantId"), ("principal", "PrincipalId"), ("descriptorDigest", "Digest"), ("expiry", "Timestamp"), ("maxUses", "IntString"), ("purposeDigest", "Digest")):
        _primitive(kind, source[field], f"CapabilityGrant/{field}")
    _strings(source["actions"], "CapabilityGrant/actions", nonempty=True)
    _selector(source["selector"], "CapabilityGrant/selector")
    constraints = _object(source["constraints"], "CapabilityGrant/constraints")
    _required(constraints, ("budgetLeaseId",), "CapabilityGrant/constraints")
    _primitive("LeaseId", constraints["budgetLeaseId"], "CapabilityGrant/constraints/budgetLeaseId")
    for field in ("maxBytes", "maxEffects"):
        if field in constraints:
            _primitive("IntString", constraints[field], f"CapabilityGrant/constraints/{field}")
    if "environmentSnapshot" in constraints:
        _primitive("Digest", constraints["environmentSnapshot"], "CapabilityGrant/constraints/environmentSnapshot")
    if "networkPolicy" in constraints:
        _enum(constraints["networkPolicy"], ("deny", "allowlist"), "CapabilityGrant/constraints/networkPolicy")
    if "requirePreview" in constraints and not isinstance(constraints["requirePreview"], bool):
        raise WireError("type", "CapabilityGrant/constraints/requirePreview", "expected boolean")
    if "requireApprovalAboveRisk" in constraints:
        _primitive("RiskTier", constraints["requireApprovalAboveRisk"], "CapabilityGrant/constraints/requireApprovalAboveRisk")
    if "parentGrantId" in source:
        _primitive("GrantId", source["parentGrantId"], "CapabilityGrant/parentGrantId")
    if "approvalRef" in source:
        _primitive("ApprovalId", source["approvalRef"], "CapabilityGrant/approvalRef")
    if "authenticator" in source:
        _string(source["authenticator"], "CapabilityGrant/authenticator")


def _parse_receipt(value: Any) -> None:
    source = _object(value, "Receipt")
    _required(source, ("descriptorDigest", "outcome", "observedAt", "resultDigest", "affectedResources"), "Receipt")
    for field, kind in (("descriptorDigest", "Digest"), ("observedAt", "Timestamp"), ("resultDigest", "Digest")):
        _primitive(kind, source[field], f"Receipt/{field}")
    outcome = _enum(source["outcome"], ("ok", "failed", "undeterminable"), "Receipt/outcome")
    if "grantId" in source:
        _primitive("GrantId", source["grantId"], "Receipt/grantId")
    for index, item in enumerate(_array(source["affectedResources"], "Receipt/affectedResources")):
        item_path = f"Receipt/affectedResources/{index}"
        record = _object(item, item_path)
        _required(record, ("resource", "change"), item_path)
        _string(record["resource"], f"{item_path}/resource")
        change = _enum(record["change"], ("created", "modified", "deleted", "observed"), f"{item_path}/change")
        if change != "created" and "preDigest" not in record:
            raise WireError("required", item_path, "preDigest required")
        if change != "deleted" and "postDigest" not in record:
            raise WireError("required", item_path, "postDigest required")
        for field in ("preDigest", "postDigest", "patchRef"):
            if field in record:
                _primitive("Digest", record[field], f"{item_path}/{field}")
    if outcome == "undeterminable":
        if "uncertainty" not in source:
            raise WireError("required", "Receipt", "undeterminable requires uncertainty")
        uncertainty = _object(source["uncertainty"], "Receipt/uncertainty")
        _required(uncertainty, ("scope", "reason"), "Receipt/uncertainty")
        _enum(uncertainty["scope"], ("effect_occurrence", "evidence_completeness", "result"), "Receipt/uncertainty/scope")
        _string(uncertainty["reason"], "Receipt/uncertainty/reason")


def _parse_event(value: Any) -> None:
    source = _object(value, "EventEnvelope")
    required = ("schemaVersion", "eventId", "scope", "traceId", "spanId", "seq", "occurredAt", "recordedAt", "principal", "principalRole", "tenantId", "ownerId", "confidentiality", "retentionClass", "trainability", "redactionStatus", "payload")
    _required(source, required, "EventEnvelope")
    if source["schemaVersion"] != "vg.4":
        raise WireError("const", "EventEnvelope/schemaVersion", "unsupported schema version")
    scope = _enum(source["scope"], ("episode", "governance", "evolution", "recovery"), "EventEnvelope/scope")
    for field, kind in (("eventId", "Uuidv7"), ("seq", "IntString"), ("occurredAt", "Timestamp"), ("recordedAt", "Timestamp"), ("principal", "PrincipalId"), ("tenantId", "TenantId"), ("ownerId", "OwnerId"), ("confidentiality", "ConfidentialityLabel"), ("retentionClass", "RetentionClass"), ("trainability", "TrainabilityLabel"), ("redactionStatus", "RedactionStatus")):
        _primitive(kind, source[field], f"EventEnvelope/{field}")
    _enum(source["principalRole"], ("user", "operator", "episode", "process", "evaluator", "release"), "EventEnvelope/principalRole")
    _string(source["traceId"], "EventEnvelope/traceId")
    _string(source["spanId"], "EventEnvelope/spanId")
    if scope in ("episode", "recovery") and "runId" not in source:
        raise WireError("required", "EventEnvelope", f"{scope} requires runId")
    if scope == "episode" and "episodeId" not in source:
        raise WireError("required", "EventEnvelope", "episode requires episodeId")
    if scope in ("governance", "evolution") and "runId" in source:
        raise WireError("scope", "EventEnvelope/runId", f"{scope} cannot carry runId")
    if scope != "episode" and "episodeId" in source:
        raise WireError("scope", "EventEnvelope/episodeId", f"{scope} cannot carry episodeId")
    payload = _object(source["payload"], "EventEnvelope/payload")
    _required(payload, ("kind",), "EventEnvelope/payload")
    _string(payload["kind"], "EventEnvelope/payload/kind")


def _parse_artifact(value: Any) -> None:
    source = _object(value, "Artifact")
    _required(source, ("artifactId", "kind", "class", "hypothesis", "evidenceRefs", "invalidationConditions", "riskDelta", "contentDigest"), "Artifact")
    _primitive("ArtifactId", source["artifactId"], "Artifact/artifactId")
    _string(source["kind"], "Artifact/kind")
    artifact_class = _enum(source["class"], ("enforcement", "compensation"), "Artifact/class")
    if artifact_class == "compensation" and "compensatesFor" not in source:
        raise WireError("required", "Artifact", "compensation requires compensatesFor")
    if artifact_class == "enforcement" and "compensatesFor" in source:
        raise WireError("not", "Artifact/compensatesFor", "enforcement cannot compensate")
    _string(source["hypothesis"], "Artifact/hypothesis")
    for index, digest in enumerate(_array(source["evidenceRefs"], "Artifact/evidenceRefs")):
        _primitive("Digest", digest, f"Artifact/evidenceRefs/{index}")
    _invalidation(source["invalidationConditions"], "Artifact/invalidationConditions")
    _integer(source["riskDelta"], "Artifact/riskDelta")
    _primitive("Digest", source["contentDigest"], "Artifact/contentDigest")
    content = {key: val for key, val in source.items() if key != "contentDigest" and key in {"artifactId", "kind", "class", "compensatesFor", "hypothesis", "evidenceRefs", "invalidationConditions", "riskDelta"}}
    if digest_of(content) != source["contentDigest"]:
        raise WireError("digest", "Artifact/contentDigest", "does not bind immutable content")


def _parse_claim(value: Any) -> None:
    source = _object(value, "EvidenceClaim")
    _required(source, ("id", "subject", "predicate", "value", "protocol", "evaluator", "environmentProfile", "substrateProfile", "taskDistribution", "uncertainty", "validity", "invalidationConditions"), "EvidenceClaim")
    for field, kind in (("id", "ClaimId"), ("protocol", "Digest"), ("environmentProfile", "Digest"), ("substrateProfile", "Digest"), ("taskDistribution", "Digest")):
        _primitive(kind, source[field], f"EvidenceClaim/{field}")
    _string(source["subject"], "EvidenceClaim/subject")
    _string(source["predicate"], "EvidenceClaim/predicate")
    evaluator = _object(source["evaluator"], "EvidenceClaim/evaluator")
    _required(evaluator, ("evaluatorId", "class", "imageDigest"), "EvidenceClaim/evaluator")
    _primitive("EvaluatorId", evaluator["evaluatorId"], "EvidenceClaim/evaluator/evaluatorId")
    _primitive("Digest", evaluator["imageDigest"], "EvidenceClaim/evaluator/imageDigest")
    validity = _object(source["validity"], "EvidenceClaim/validity")
    _required(validity, ("domains",), "EvidenceClaim/validity")
    _strings(validity["domains"], "EvidenceClaim/validity/domains", nonempty=True)
    _invalidation(source["invalidationConditions"], "EvidenceClaim/invalidationConditions")


def _parse_correction(value: Any) -> None:
    source = _object(value, "CorrectionRecord")
    _required(source, ("episodeId", "proposedPatchDigest", "acceptedPatchDigest", "reasonCodes", "magnitude", "scope", "correctingPrincipalRole"), "CorrectionRecord")
    for field, kind in (("episodeId", "EpisodeId"), ("proposedPatchDigest", "Digest"), ("acceptedPatchDigest", "Digest")):
        _primitive(kind, source[field], f"CorrectionRecord/{field}")
    reasons = _strings(source["reasonCodes"], "CorrectionRecord/reasonCodes", nonempty=True)
    allowed = {"functional_defect", "missing_requirement", "security_policy", "test_inadequacy", "maintainability", "architecture_preference", "style", "product_change", "environment_change", "reviewer_disagreement"}
    if not set(reasons) <= allowed:
        raise WireError("enum", "CorrectionRecord/reasonCodes", "unknown reason code")
    _enum(source["magnitude"], ("minor", "moderate", "major"), "CorrectionRecord/magnitude")
    scope = _enum(source["scope"], ("user", "team", "repo", "domain", "general"), "CorrectionRecord/scope")
    if set(reasons) & {"style", "architecture_preference"} and scope not in {"user", "team", "repo"}:
        raise WireError("scope", "CorrectionRecord/scope", "style and preference corrections must remain local")
    _enum(source["correctingPrincipalRole"], ("user", "operator", "episode", "process", "evaluator", "release"), "CorrectionRecord/correctingPrincipalRole")


def _parse_recording(value: Any) -> None:
    source = _object(value, "Recording")
    _required(source, ("modelCassetteDigest", "imageDigest", "envSnapshotDigest", "seed", "clockPolicy"), "Recording")
    for field in ("modelCassetteDigest", "imageDigest", "envSnapshotDigest"):
        _primitive("Digest", source[field], f"Recording/{field}")
    _primitive("IntString", source["seed"], "Recording/seed")
    _enum(source["clockPolicy"], ("recorded", "fixed", "logical"), "Recording/clockPolicy")


def _parse_process_definition(value: Any) -> None:
    source = _object(value, "ProcessDefinition")
    _required(source, ("definitionDigest", "states", "initialState", "transitions", "approvalPoints", "boundEffectVerbs"), "ProcessDefinition")
    _primitive("Digest", source["definitionDigest"], "ProcessDefinition/definitionDigest")
    states = set(_strings(source["states"], "ProcessDefinition/states", nonempty=True))
    initial = _string(source["initialState"], "ProcessDefinition/initialState")
    if initial not in states:
        raise WireError("state", "ProcessDefinition/initialState", "initial state is undeclared")
    transitions = _array(source["transitions"], "ProcessDefinition/transitions")
    seen: set[tuple[str, str]] = set()
    for index, value in enumerate(transitions):
        path = f"ProcessDefinition/transitions/{index}"
        transition = _object(value, path)
        _required(transition, ("from", "eventKind", "to"), path)
        start, event, end = (_string(transition[key], f"{path}/{key}") for key in ("from", "eventKind", "to"))
        if start not in states or end not in states:
            raise WireError("state", path, "transition references undeclared state")
        if (start, event) in seen:
            raise WireError("conflict", path, "transition is nondeterministic")
        seen.add((start, event))
    _strings(source["approvalPoints"], "ProcessDefinition/approvalPoints")
    _strings(source["boundEffectVerbs"], "ProcessDefinition/boundEffectVerbs")
    content = {key: source[key] for key in ("states", "initialState", "transitions", "approvalPoints", "boundEffectVerbs")}
    if digest_of(content) != source["definitionDigest"]:
        raise WireError("digest", "ProcessDefinition/definitionDigest", "does not bind definition")


def _parse_process_instance(value: Any) -> None:
    source = _object(value, "ProcessInstance")
    _required(source, ("processId", "definitionDigest", "currentState", "allowedTransitions", "pendingApprovals", "boundEffectVerbs", "history"), "ProcessInstance")
    _primitive("ProcessId", source["processId"], "ProcessInstance/processId")
    _primitive("Digest", source["definitionDigest"], "ProcessInstance/definitionDigest")
    _string(source["currentState"], "ProcessInstance/currentState")
    _strings(source["allowedTransitions"], "ProcessInstance/allowedTransitions")
    for index, approval in enumerate(_array(source["pendingApprovals"], "ProcessInstance/pendingApprovals")):
        _primitive("ApprovalId", approval, f"ProcessInstance/pendingApprovals/{index}")
    _strings(source["boundEffectVerbs"], "ProcessInstance/boundEffectVerbs")
    for index, value in enumerate(_array(source["history"], "ProcessInstance/history")):
        path = f"ProcessInstance/history/{index}"
        record = _object(value, path)
        _required(record, ("from", "eventKind", "to", "eventId"), path)
        for field in ("from", "eventKind", "to"):
            _string(record[field], f"{path}/{field}")
        _primitive("Uuidv7", record["eventId"], f"{path}/eventId")


_PARSERS: dict[str, Callable[[Any], None]] = {
    "EffectDescriptor": _parse_effect,
    "CapabilityGrant": _parse_grant,
    "Receipt": _parse_receipt,
    "EventEnvelope": _parse_event,
    "Artifact": _parse_artifact,
    "EvidenceClaim": _parse_claim,
    "CorrectionRecord": _parse_correction,
    "Recording": _parse_recording,
    "ProcessDefinition": _parse_process_definition,
    "ProcessInstance": _parse_process_instance,
}
WIRE_KINDS = tuple(_PARSERS)


def parse_wire(kind: str, value: Any) -> JsonObject:
    """Validate one reader-profile value and return a lossless deep copy."""

    parser = _PARSERS.get(kind)
    if parser is None:
        raise WireError("kind", kind, "unknown wire contract")
    parser(value)
    return deepcopy(_object(value, kind))
